File: src/main/java/io/jenkins/plugins/gamekins/challenge/TestChallenge.java
package io.jenkins.plugins.gamekins.challenge;

import hudson.FilePath;
import hudson.model.Run;
import hudson.model.TaskListener;
import hudson.model.User;
import io.jenkins.plugins.gamekins.util.GitUtil;
import io.jenkins.plugins.gamekins.util.JacocoUtil;
import org.jenkinsci.plugins.workflow.job.WorkflowJob;
import org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject;

import java.io.IOException;
import java.util.HashMap;
import java.util.Set;

public class TestChallenge implements Challenge {

    private final String currentCommit;
    private final int testCount;
    private final User user;
    private final String branch;
    private final long created = System.currentTimeMillis();
    private long solved = 0;
    private int testCountSolved = 0;

    public TestChallenge(String currentCommit, int testCount, User user, String branch) {
        this.currentCommit = currentCommit;
        this.testCount = testCount;
        this.user = user;
        this.branch = branch;
    }

    @Override
    public boolean isSolved(HashMap<String, String> constants, Run<?, ?> run, TaskListener listener,
                            FilePath workspace) {
        if (!this.branch.equals(constants.get("branch"))) return false;
        try {
            int testCountSolved = JacocoUtil.getTestCount(workspace, run);
            if (testCountSolved <= this.testCount) {
                return false;
            }

            Set<String> lastChangedFilesOfUser = GitUtil.getLastChangedTestFilesOfUser(
                            workspace, user, 0, currentCommit, User.getAll());
            if (lastChangedFilesOfUser.size() > 0) {
                this.solved = System.currentTimeMillis();
                this.testCountSolved = testCountSolved;
                return true;
            }
        } catch (IOException | InterruptedException e) {
            e.printStackTrace(listener.getLogger());
        }
        return false;
    }

    @Override
    public boolean isSolvable(HashMap<String, String> constants, Run<?, ?> run, TaskListener listener,
                              FilePath workspace) {
        if (run.getParent().getParent() instanceof WorkflowMultiBranchProject) {
            for (WorkflowJob workflowJob : ((WorkflowMultiBranchProject) run.getParent().getParent()).getItems()) {
                if (workflowJob.getName().equals(this.branch)) return true;
            }
        } else {
            return true;
        }
        return false;
    }

    @Override
    public int getScore() {
        return 1;
    }

    @Override
    public long getCreated() {
        return this.created;
    }

    @Override
    public long getSolved() {
        return this.solved;
    }

    @Override
    public String printToXML(String reason, String indentation) {
        String print = indentation + "<TestChallenge created=\"" + this.created + "\" solved=\"" + this.solved
                + "\" tests=\"" + this.testCount + "\" testsAtSolved=\"" + this.testCountSolved;
        if (!reason.isEmpty()) {
            print += "\" reason=\"" + reason;
        }
        print += "\"/>";
        return print;
    }

    @Override
    public String toString() {
        return "Write a new test in branch " + branch;
    }
}

File: src/main/kotlin/io/jenkins/plugins/gamekins/challenge/TestChallenge.kt
package io.jenkins.plugins.gamekins.challenge

import hudson.FilePath
import hudson.model.Run
import hudson.model.TaskListener
import hudson.model.User
import io.jenkins.plugins.gamekins.util.GitUtil.getLastChangedTestFilesOfUser
import io.jenkins.plugins.gamekins.util.JacocoUtil.getTestCount
import org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject
import java.util.*

class TestChallenge(private val currentCommit: String, private val testCount: Int, private val user: User,
                    private val branch: String) : Challenge {

    private val created = System.currentTimeMillis()
    private var solved: Long = 0
    private var testCountSolved = 0

    override fun getCreated(): Long {
        return created
    }

    override fun getScore(): Int {
        return 1
    }

    override fun getSolved(): Long {
        return solved
    }

    override fun isSolvable(constants: HashMap<String, String>, run: Run<*, *>, listener: TaskListener,
                            workspace: FilePath): Boolean {
        if (run.parent.parent is WorkflowMultiBranchProject) {
            for (workflowJob in (run.parent.parent as WorkflowMultiBranchProject).items) {
                if (workflowJob.name == branch) return true
            }
        } else {
            return true
        }
        return false
    }

    override fun isSolved(constants: HashMap<String, String>, run: Run<*, *>, listener: TaskListener,
                          workspace: FilePath): Boolean {
        if (branch != constants["branch"]) return false
        try {
            val testCountSolved = getTestCount(workspace, run)
            if (testCountSolved <= testCount) {
                return false
            }
            val lastChangedFilesOfUser = getLastChangedTestFilesOfUser(
                    workspace, user, 0, currentCommit, User.getAll())
            if (lastChangedFilesOfUser.isNotEmpty()) {
                solved = System.currentTimeMillis()
                this.testCountSolved = testCountSolved
                return true
            }
        } catch (e: Exception) {
            e.printStackTrace(listener.logger)
        }
        return false
    }

    override fun printToXML(reason: String, indentation: String): String {
        var print = (indentation + "<TestChallenge created=\"" + created + "\" solved=\"" + solved
                + "\" tests=\"" + testCount + "\" testsAtSolved=\"" + testCountSolved)
        if (reason.isNotEmpty()) {
            print += "\" reason=\"$reason"
        }
        print += "\"/>"
        return print
    }

    override fun toString(): String {
        return "Write a new test in branch $branch"
    }
}
